### ISSUE TITLE
Support globs with --test arg

### ISSUE THREAD

**Issue by humphd** [Orange-OpenSource/hurl#387]
I'm starting to get a lot of `.hurl` test files, and I'd like to be able to split them up into different directories.  Using `--test` I can specify `dir/*` and it will happily tear through all the `.hurl` files one level deep.  However, if I have a directory in there, I get:

```sh
error: Is a directory (os error 21)
```

It would be great if I could use `--test dir/**/*.hurl` or something to have it descend into sub directories.

**Comment by fabricereix** [comment]
Hello @humphd,
Note that Hurl does not process the '*' as such.
From the command line `hurl tests/*.hurl`, the terminal carries out the filename expansion.
Hurl then simply processes the following command-line `hurl tests/test1.hurl tests/test2.hurl tests/test3.hurl .

Note that `--test` run Hurl in "test"mode , turning on progress and summary.
It is orthogonal to the hurl input files, that's why I've removed from your examples.

For recursive globing, the following command-line can be used:
`hurl $(find tests -name *.hurl)`

Relating to adding recursive globbing in Hurl, we could add an explicit `--glob` option, like ripgrep.
`hurl --glob "*.hurl"`

**Comment by humphd** [comment]
Thanks for the note about `--test`.  One of the reasons I'm interested in having a way to do this using only Hurl is that I'd like to have my students use it, and they are predominately on Windows, where using `find` isn't going to work.  I understand not wanting to expand the scope of Hurl too far, but I think ingesting trees of `.hurl` files seems in scope.  Your `--glob` approach would be great.

**Comment by jcamiel** [comment]
@humphd we're going to implement `--glob` in the next Hurl version (1.5.0): indeed it's a good idea especially for Windows. If you plan to use it with students some day, I'm really curious about your feedbacks: we use Hurl intensively in-house, but we look forward to have more external feedbacks!

**Comment by jcamiel** [comment]
Really interesting, thank you! I wish I had this kind of CS courses !!

**Comment by humphd** [comment]
> I wish I had this kind of CS courses !!

Me too! That's why I do it. CS was nothing like this when I went through.

**Comment by fabricereix** [comment]
Here are a few examples to illustrate how the `--glob` option can fit into Hurl.

Assuming that the Hurl input files are:

```
$ tree
.
└── tests
    ├── project1
    │   ├── test1.hurl
    │   └── test2.hurl
    └── project2
        └── test3.hurl

```

### Without the glob option

Recursive globbing is supported in zsh
```
% ls **/*.hurl
tests/project1/test1.hurl  tests/project2/test3.hurl
tests/project1/test2.hurl

% hurl **/*.hurl
# will execute the command hurl tests/project1/test1.hurl tests/project1/test2.hurl tests/project2/test3.hurl
```

While it is not in bash

```
ls **/*.hurl
ls: cannot access '**/*.hurl': No such file or directory

$ hurl **/*.hurl
error: No such file or directory (os error 2)
```


### Using the `--glob` option 

```
hurl --glob **/*.hurl
```

The 3 Hurl files are executing in both bash and zsh


###  Combining glob and file expansion in the shell

for example
```
hurl --glob tests/**/*.hurl
```

It works as expected the 3 Hurl files are run.
But under the hood, this is different from the previous exemple.

In this case, the files are expanded first by the shell

```
ls tests/**/*.hurl
tests/project1/test1.hurl  tests/project2/test3.hurl
tests/project1/test2.hurl
```

The command run is 
```
hurl --glob tests/project1/test1.hurl tests/project2/test3.hurl tests/project1/test2.hurl
```

The glob expression is therfore `tests/project1/test1.hurl`
and the "standard" input files `are tests/project1/test2.hurl and `tests/project1/test2.hurl`.


It also "works" when adding other options after the glob option.
Because options can be specified (luckilly) after the input files.

```
hurl --glob tests/**/*.hurl -v
```
is the same as 
```hurl --glob tests/project1/test1.hurl tests/project1/test2.hurl tests/project2/test3.hurl -v```

### Using globbing only from Hurl

You can also prevents the shell expansion by encapsulating the glob expression between quotes.

```
hurl --glob "tests/**/*.hurl"
```

In this case, the 3 Hurl files are only from the glob in Hurl, and there is no input files given at the command-line.


### Running hurl without input files 

It prints the usage and exits.
This is still the case with the glob option: at least one file must be provided, either as input file at the command-line or produced from the glob expression.

**Comment by humphd** [comment]
> I teach software development, open source, and web/cloud to post-secondary students in Toronto, Canada. For an upcoming course on AWS and microservices, I want to have students write integration tests against their containers. Because they all usually work in node.js, I have tended to have them use pure JS tooling for this (e.g., Jest, mocha, Supertest, etc). I still do this for unit testing.

A brief follow-up on this, and word of encouragement.

Next week is my final week teaching the course I mentioned above, and I wanted to report back to let you know that Hurl has been wonderful to use with the students.  I spent a lecture showing them how it worked, and how to write integration tests with it against Docker containers and AWS services.  They picked it up really quickly, and were able to use it for the subsequent weeks without much help or guidance.  I asked them what their experience was like, and all of them found it quick to learn and fast to write.  It caught many bugs, including lots of cases where students forget to `await` a `Promise`, and you often miss the race unless you have a test runner as fast as Hurl.  One of my students [wrote a blog post about his experience (second-half of this post)](https://dev.to/pandanoxes/discovered-new-tools-29al).

I wanted to say 'thanks' for all your support with the various issues and requests I've filed.  I'm teaching the course to a larger group in May, so we'll have a chance to do it all again.

Thanks again, and keep up the great work!
